cpu/fe310/periph/uart.c

```diff
@@ -86,77 +83,51 @@ int uart_init(uart_t dev, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
-            UART0_REG(UART_REG_IE) = UART_IP_RXWM;
-            UART0_REG(UART_REG_RXCTRL) = UART_RXEN;
-
-            /* Re-eanble ext interrupts */
-            set_csr(mie, MIP_MEIP);
-        }
-    }
+    _REG32(uart_config[dev].addr, UART_REG_DIV) = uartDiv;
 
-    if (dev == 1) {
-        /* Config UART */
-        UART1_REG(UART_REG_DIV) = uartDiv;
-        UART1_REG(UART_REG_TXCTRL) = UART_TXEN;
-
-        /* Enable RX intr if there is a callback */
-        if (rx_cb != NULL) {
-            /* Disable ext interrupts when setting up */
-            clear_csr(mie, MIP_MEIP);
-
-            /* Configure UART ISR with PLIC */
-            set_external_isr_cb(INT_UART1_BASE, uart_isr);
-            PLIC_enable_interrupt(INT_UART1_BASE);
-            PLIC_set_priority(INT_UART1_BASE, UART1_RX_INTR_PRIORITY);
-            UART1_REG(UART_REG_IE) = UART_IP_RXWM;
-            UART1_REG(UART_REG_RXCTRL) = UART_RXEN;
-
-            /* Re-eanble ext interrupts */
-            set_csr(mie, MIP_MEIP);
-        }
+    /* Select IOF0 */
+    GPIO_REG(GPIO_IOF_SEL) &= ~(1 << uart_config[dev].tx);
+    /* Enable IOF */
+    GPIO_REG(GPIO_IOF_EN) |= (1 << uart_config[dev].tx);
+
+    /* Enable TX */
+    _REG32(uart_config[dev].addr, UART_REG_TXCTRL) = UART_TXEN;
+
+    /* Enable RX intr if there is a callback */
+    if (rx_cb) {
+        /* Select IOF0 */
+        GPIO_REG(GPIO_IOF_SEL) &= ~(1 << uart_config[dev].rx);
+        /* Enable IOF */
+        GPIO_REG(GPIO_IOF_EN) |= (1 << uart_config[dev].rx);
+
+        /* Disable ext interrupts when setting up */
+        clear_csr(mie, MIP_MEIP);
+
+        /* Configure UART ISR with PLIC */
+        set_external_isr_cb(uart_config[dev].isr_num, uart_isr);
+        PLIC_enable_interrupt(uart_config[dev].isr_num);
+        PLIC_set_priority(uart_config[dev].isr_num, UART_ISR_PRIO);
+        _REG32(uart_config[dev].addr, UART_REG_IE) = UART_IP_RXWM;
+
+        /* Enable RX */
+        _REG32(uart_config[dev].addr, UART_REG_RXCTRL) = UART_RXEN;
+
+        /* Re-enable ext interrupts */
+        set_csr(mie, MIP_MEIP);
     }
 
     return UART_OK;
 }
 
 void uart_write(uart_t dev, const uint8_t *data, size_t len)
 {
-    if (dev == 0) {
-        for (size_t i = 0; i < len; i++) {
-            /* Wait for FIFO to empty */
-            while ((UART0_REG(UART_REG_TXFIFO) & UART_TXFIFO_FULL)
-                   == (uint32_t)UART_TXFIFO_FULL)
-                ;
-
-            /* Write a byte */
-            UART0_REG(UART_REG_TXFIFO) = data[i];
-        }
-    }
-
-    if (dev == 1) {
-        for (size_t i = 0; i < len; i++) {
-            /* Wait for FIFO to empty */
-            while ((UART1_REG(UART_REG_TXFIFO) & UART_TXFIFO_FULL)
-                   == (uint32_t)UART_TXFIFO_FULL)
-                ;
+    for (size_t i = 0; i < len; i++) {
+        /* Wait for FIFO to empty */
+        while ((_REG32(uart_config[dev].addr, UART_REG_TXFIFO) & UART_TXFIFO_FULL)
+               == (uint32_t)UART_TXFIFO_FULL) {};
```

Comment:
Is this cast needed?

Reply:
Yes, it doesn't build without (error: comparison of integer expressions of different signedness"

Reply:
OK!